engine/src/main/java/org/pentaho/di/trans/Trans.java

```diff
@@ -0,0 +1,5804 @@
+        //
+        PerformanceLogTable performanceLogTable = transMeta.getPerformanceLogTable();
+        int perfLogInterval = Const.toInt( environmentSubstitute( performanceLogTable.getLogInterval() ), -1 );
+        if ( performanceLogTable.isDefined() && perfLogInterval > 0 ) {
+          final Timer timer = new Timer( getName() + " - step performance log interval timer" );
+          TimerTask timerTask = new TimerTask() {
+            @Override
+            public void run() {
+              try {
+                lastWrittenStepPerformanceSequenceNr =
+                  writeStepPerformanceLogRecords( lastWrittenStepPerformanceSequenceNr, LogStatus.RUNNING );
+              } catch ( Exception e ) {
+                log.logError( BaseMessages.getString( PKG,
+                  "Trans.Exception.UnableToPerformIntervalPerformanceLogging" ), e );
+                // Also stop the show...
+                //
+                errors.incrementAndGet();
+                stopAll();
+              }
+            }
+          };
+          timer.schedule( timerTask, perfLogInterval * 1000, perfLogInterval * 1000 );
+
+          addTransListener( new TransAdapter() {
+            @Override
+            public void transFinished( Trans trans ) {
+              timer.cancel();
+            }
+          } );
+        }
+      } catch ( KettleException e ) {
+        throw new KettleTransException( BaseMessages.getString( PKG, "Trans.Exception.ErrorWritingLogRecordToTable",
+          logTable ), e );
+      } finally {
+        // If we use interval logging, we keep the connection open for performance reasons...
+        //
+        if ( transLogTableDatabaseConnection != null && ( intervalInSeconds <= 0 ) ) {
+          transLogTableDatabaseConnection.disconnect();
+          transLogTableDatabaseConnection = null;
+        }
+      }
+    } catch ( KettleException e ) {
+      throw new KettleTransException( BaseMessages.getString( PKG,
+        "Trans.Exception.UnableToBeginProcessingTransformation" ), e );
+    }
+  }
+
+  /**
+   * Writes log channel information to a channel logging table (if one has been configured).
+   *
+   * @throws KettleException if any errors occur during logging
+   */
+  protected void writeLogChannelInformation() throws KettleException {
+    Database db = null;
+    ChannelLogTable channelLogTable = transMeta.getChannelLogTable();
+
+    // PDI-7070: If parent trans or job has the same channel logging info, don't duplicate log entries
+    Trans t = getParentTrans();
+    if ( t != null ) {
+      if ( channelLogTable.equals( t.getTransMeta().getChannelLogTable() ) ) {
+        return;
+      }
+    }
+
+    Job j = getParentJob();
+
+    if ( j != null ) {
+      if ( channelLogTable.equals( j.getJobMeta().getChannelLogTable() ) ) {
+        return;
+      }
+    }
+    // end PDI-7070
+
+    try {
+      db = new Database( this, channelLogTable.getDatabaseMeta() );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *db*. The resource is referenced at lines 2595, 2596, 2597, 2601, 2606, 2611. The resource is closed at line 2722. There are other execution paths that do not contain closure statements, for example, when Database.commit (in method disconnectDb) throws an exception, and an if condition at line 2719 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).